### DISCUSSION TITLE
Dark Mode implementation production example

### DISCUSSION THREAD

**Discussion by chmcquee** [keycloakify/keycloakify#248] [Q&A] [answered]
Hi @garronej  
I'm now focused on making the login screen available in darkMode.
For this I read your docs.

However there was some confusion what file to paste the following code into

```tsx 
import keycloak_js from "keycloak-js";
import { injectGlobalStatesInSearchParams } from "powerhooks/useGlobalState";
import { createKeycloakAdapter } from "keycloakify";
import { addParamToUrl } from "powerhooks/tools/urlSearchParams";

//...

const keycloakInstance = keycloak_js({
    "url": "http://keycloak-server/auth",
    "realm": "myrealm",
    "clientId": "myapp",
});

keycloakInstance.init({
    "onLoad": "check-sso",
    "silentCheckSsoRedirectUri": window.location.origin + "/silent-check-sso.html",
    "adapter": createKeycloakAdapter({
        "transformUrlBeforeRedirect": url=>
            [url]
                //This will append &ui_locales=fr at on the login page url we are about to 
                //redirect to. 
                //This will tell keycloak that the login should be in french. 
                //Replace "fr" by any KcLanguageTag you have enabled on your Keycloak server.
                .map(url => addParamToUrl({ url, "name": "ui_locales", "value": "fr" }).newUrl)
                //This will make sure our isDarkModeEnabled and other global states
                //created with powerhooks/useGlobalState are restored on the other side. 
                .map(injectGlobalStatesInSearchParams)
                [0],
        keycloakInstance,
    }),
});

//...
```

I'd really appreciate some more clarity. Also if you had a darkmode production app live, that'd be very helpful. 

Thanks again,
Chris

**Comment by garronej** [comment, marked as answer]
Hi @chmcquee,  
What are you trying to acheive?  
Are you trying to create a login page for a React App that already exists or are you creating a standalone Keycloak theme?  
If you are creating a standalone Keycloak theme you can implement a dark mode mechanism however you see fit.  
Now the documentation you are refering to is to consider if you wish that the dark mode state be carried away from your main app to the login pages.   
That is to say, the user is using dark mode, he navigates to the login page, we want the dark mode to be persisted.  
However this is a small UX detail, you can very well use the OS preferred default and it will work in 99% of the case.  

This App: https://datalab.sspcloud.fr/ has dark mode state caried over to the login pages, the code is here https://github.com/InseeFrLab/onyxia-web.  
But I doubt it would be of any help to you since this implementation is very sophisticated.

**Reply by chmcquee** [reply]
Thanks for your reply @garronej 
I figured out that the code referenced in my comment refers to the main app setup for keycloak, which we do have in our project.
I was also wondering if you had could help a little more by explaining [this example](https://github.com/garronej/powerhooks/blob/main/src/test/spa/src/TestUseGlobalState/useIsDarkModeEnabled.tsx) with some comments as it's not 100% clear to me what next and subscribe do.

**Reply by garronej** [reply]
I added comments.  
If you use [this example](https://github.com/garronej/powerhooks/blob/main/src/test/spa/src/TestUseGlobalState/useIsDarkModeEnabled.tsx) and use [`injectGlobalStatesInSearchParams`](https://docs.keycloakify.dev/context-persistence) things should work trensparently.

**Reply by chmcquee** [reply]
Thanks so much @garronej . I've found a solution that works on class level. Appreciate your effort!
